### PR TITLE
MINIFICPP-792 - TailFile processor should handle rotation of source file

### PR DESCRIPTION
**Pull request by arpadboda**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: libminifi/src/processors/TailFile.cpp [hunk 1]
```diff
@@ -179,58 +180,22 @@ void TailFile::checkRollOver(const std::string &fileLocation, const std::string
...
-            TailMatchedFileItem item;
-            item.fileName = fileName;
-            item.modifiedTime = ((uint64_t) (statbuf.st_mtime) * 1000);
-            matchedFiles.push_back(item);
-          }
+
+    auto lambda = [&](const std::string& path, const std::string& filename) -> bool {
+      struct stat sb;
+      std::string fileFullName = path + utils::file::FileUtils::get_separator() + filename;
+      if (fileFullName.find(pattern) != std::string::npos && stat(fileFullName.c_str(), &sb) == 0) {
```
Comment (arpadboda) [inline libminifi/src/processors/TailFile.cpp]: EL support not added yet, this is the more important part (fixing bugs), I will add that as a separate PR.

File: libminifi/src/processors/TailFile.cpp [hunk 2]
```diff
@@ -36,6 +36,7 @@
 #include <sstream>
 #include <string>
 #include <iostream>
+#include <utils/file/FileUtils.h>
```
Comment (phrocker) [inline libminifi/src/processors/TailFile.cpp]: We should keep with the current formatting. I would prefer a change too but we should probably avoid making stylistic changes.

File: libminifi/src/processors/PutFile.cpp [hunk 3]
```diff
@@ -138,49 +138,18 @@ void PutFile::onTrigger(core::ProcessContext *context, core::ProcessSession *ses
     if (S_ISDIR(statResult.st_mode)) {
       // it's a directory, count the files
       int64_t ct = 0;
-#ifndef WIN32
-      DIR *myDir = opendir(directory.c_str());
-      if (!myDir) {
-        logger_->log_warn("Could not open %s", directory);
+      auto lambda = [&ct, this](const std::string&, const std::string&) -> bool {
```
Comment (phrocker) [inline libminifi/src/processors/PutFile.cpp]: If we're going to use lambdas here we need a comment here about what the return means for the follow on function.

File: libminifi/src/processors/TailFile.cpp (hunk 1 above)
Comment (phrocker) [reply libminifi/src/processors/TailFile.cpp]: The ticket is labeled an improvement. If it is a bug fix we should isolate this to TailFile.

File: libminifi/src/processors/TailFile.cpp [hunk 4]
```diff
@@ -179,58 +180,22 @@ void TailFile::checkRollOver(const std::string &fileLocation, const std::string
...
-          }
-        }
-      }while (FindNextFile(hFind, &FindFileData));
-      FindClose(hFind);
-    }
-#endif
+    utils::file::FileUtils::list_dir(fileLocation, lambda, logger_, false);
 
     // Sort the list based on modified time
     std::sort(matchedFiles.begin(), matchedFiles.end(), sortTailMatchedFileItem);
```
Comment (phrocker) [inline libminifi/src/processors/TailFile.cpp]: Sorting is probably invalid now

File: libminifi/src/processors/TailFile.cpp [hunk 5]
```diff
@@ -179,58 +180,22 @@ void TailFile::checkRollOver(const std::string &fileLocation, const std::string
...
-        std::string fileFullName = fileLocation + "/" + d_name;
-        if (fileFullName.find(pattern) != std::string::npos && stat(fileFullName.c_str(), &statbuf) == 0) {
-          if (((uint64_t) (statbuf.st_mtime) * 1000) >= modifiedTimeCurrentTailFile) {
-            TailMatchedFileItem item;
-            item.fileName = fileName;
-            item.modifiedTime = ((uint64_t) (statbuf.st_mtime) * 1000);
-            matchedFiles.push_back(item);
-          }
+
+    auto lambda = [&](const std::string& path, const std::string& filename) -> bool {
```
Comment (phrocker) [inline libminifi/src/processors/TailFile.cpp]: same about commenting this.

File: libminifi/src/processors/TailFile.cpp (hunk 1 above)
Comment (arpadboda) [reply libminifi/src/processors/TailFile.cpp]: Changed the ticket to "bug".
When I created the ticket I thought the feature is completely missing, although I found it, just not working as expected.

File: libminifi/src/processors/TailFile.cpp (hunk 4 above)
Comment (arpadboda) [reply libminifi/src/processors/TailFile.cpp]: Why would it be? The content of the container should be the same.

File: libminifi/src/processors/TailFile.cpp (hunk 2 above)
Comment (arpadboda) [reply libminifi/src/processors/TailFile.cpp]: Good spot, thanks! I was my IDE doing it automatically. Fixed.

File: libminifi/src/processors/TailFile.cpp (hunk 5 above)
Comment (arpadboda) [reply libminifi/src/processors/TailFile.cpp]: Added comment

File: libminifi/src/processors/PutFile.cpp (hunk 3 above)
Comment (arpadboda) [reply libminifi/src/processors/PutFile.cpp]: Added.

File: libminifi/src/processors/TailFile.cpp (hunk 1 above)
Comment (phrocker) [reply libminifi/src/processors/TailFile.cpp]: Can you remove the Putfile impl? That will help with any future git activity on the bug fix.

File: libminifi/src/processors/TailFile.cpp (hunk 4 above)
Comment (phrocker) [reply libminifi/src/processors/TailFile.cpp]: As per the java implementation if the times are the same we want to always we interrogate those rollovers properly, but i'm going to take a look a bit closer at this PR. 

What tests did you perform to validate this? Did you have a TailFileTest for me to validate against?

Comment (arpadboda) [reply libminifi/src/processors/TailFile.cpp]:
```
MiNiFi Config Version: 3
Flow Controller:
  name: techops
  comment: Created by MiNiFi C2 Flow Designer
Core Properties:
  flow controller graceful shutdown period: 10 sec
  flow service write delay interval: 500 ms
  administrative yield duration: 30 sec
  bored yield duration: 10 millis
  max concurrent threads: 1
  variable registry properties: ''
FlowFile Repository:
  partitions: 256
  checkpoint interval: 2 mins
  always sync: false
  Swap:
    threshold: 20000
    in period: 5 sec
    in threads: 1
    out period: 5 sec
    out threads: 4
Content Repository:
  content claim max appendable size: 10 MB
  content claim max flow files: 100
  always sync: false
Provenance Repository:
  provenance rollover time: 1 min
  implementation: org.apache.nifi.provenance.MiNiFiPersistentProvenanceRepository
Component Status Repository:
  buffer size: 1440
  snapshot frequency: 1 min
Security Properties:
  keystore: ''
  keystore type: ''
  keystore password: ''
  key password: ''
  truststore: ''
  truststore type: ''
  truststore password: ''
  ssl protocol: ''
  Sensitive Props:
    key:
    algorithm: PBEWITHMD5AND256BITAES-CBC-OPENSSL
    provider: BC
Processors:
- id: cb18aee6-b107-4603-98a3-689b515b2b10
  name: TailFile
  class: class org.apache.nifi.minifi.processors.TailFile
  max concurrent tasks: 1
  scheduling strategy: TIMER_DRIVEN
  scheduling period: 1000 ms
  penalization period: 30000 ms
  yield period: 1000 ms
  run duration nanos: 0
  auto-terminated relationships list: []
  Properties:
    File to Tail: /Users/aboda/work/shadow/minifi/nifi-minifi-cpp/logs/minifi-app.log
    State File: TailFileState
Controller Services: []
Process Groups: []
Input Ports: []
Output Ports: []
Funnels: []
Connections:
- id: fa777b72-6263-41a0-ba9c-7f57e0467582
  name: TailFile/success/c88192f5-0169-1000-d577-47f8091af222
  source id: cb18aee6-b107-4603-98a3-689b515b2b10
  source relationship names:
  - success
  destination id: 34948508-0166-1000-7a4c-194e61494bb1
  max work queue size: 0
  max work queue data size: 10000 B
  flowfile expiration: 60 seconds
  queue prioritizer class: ''
Remote Process Groups:
- id: 34762af9-0166-1000-1209-2a4e139b258b
  name: http://localhost:8080/nifi
  url: http://localhost:8080/nifi
  comment: ''
  timeout: 30 secs
  yield period: 10 sec
  transport protocol: RAW
  proxy host: ''
  proxy port: ''
  proxy user: ''
  proxy password: ''
  local network interface: ''
  Input Ports:
  - id: 34948508-0166-1000-7a4c-194e61494bb1
    name: MiNiFiData
    comment: ''
    max concurrent tasks: 1
    use compression: true
  Output Ports: []
NiFi Properties Overrides: {}
```

In nutshell, this tails minifi's own logfile and sends flowfile to NiFi. 
In order to verify the change I also edited minifi-log.properties file to reduce the max log file size, so rotation occurs more frequently.

File: libminifi/src/processors/TailFile.cpp (hunk 1 above)
Comment (arpadboda) [reply libminifi/src/processors/TailFile.cpp]: Done

File: libminifi/src/processors/TailFile.cpp (hunk 4 above)
Comment (phrocker) [reply libminifi/src/processors/TailFile.cpp]: Can this be added to the PR for unit tests, integration, or docker integration tests?

**Comment by arpadboda** [comment]
Amended the PR, changes done:
-Added unit tests 
-Made the behaviour compatible with NiFi: MiNiFi being stopped for a while starts to process previous files rolled over. 
-Fixed a previous issue: in case the file is not just rolled over between two "onTrigger" calls, but also written before rollover, the processor didn't handle that, just started reading the new file.

File: extensions/standard-processors/tests/unit/TailFileTests.cpp [hunk 6]
```diff
@@ -113,9 +114,175 @@ TEST_CASE("TailFileWithOutDelimiter", "[tailfiletest2]") {
...
+      tail_file,
+      processors::TailFile::StateFile.getName(), state_file);
+  auto log_attr = plan->addProcessor(
+      "LogAttribute",
+      "Log",
+      core::Relationship("success", "description"),
+      true);
+  plan->setProperty(
+      log_attr,
+      processors::LogAttribute::FlowFilesToLog.getName(), "0");
```
Comment (arpadboda) [inline extensions/standard-processors/tests/unit/TailFileTests.cpp]: @phrocker you deserve some good IPAs for this, it's so cool to test with! 🍺

**Comment by phrocker** [comment]
@arpadboda I messed up the merge ( forgot the signoff and closure statement ), but this was merged on May 30. Will close this as a result of it being merged.

**Pull request closed without merging**